File: implementations/java/segmentsort.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

public class SegmentSort {

    // Node class for the heap, storing value and segment tracking info
    private static class HeapNode {
        int value;
        int segIndex;
        int currentPos;
        int direction;

        HeapNode(int value, int segIndex, int currentPos, int direction) {
            this.value = value;
            this.segIndex = segIndex;
            this.currentPos = currentPos;
            this.direction = direction;
        }
    }

    // Segment class to store start and end indices
    private static class Segment {
        int start;
        int end;

        Segment(int start, int end) {
            this.start = start;
            this.end = end;
        }
    }

    /**
     * Sorts a list of integers using the Segment Sort algorithm.
     * @param arr The list of integers to sort.
     */
    public void sort(List<Integer> arr) {
        int n = arr.size();
        if (n <= 1) {
            return;
        }

        // Phase 1: Segment Detection
        List<Segment> segments = detectSegments(arr);

        // Phase 2: Heap Merging
        if (segments.isEmpty()) {
            return;
        }

        mergeSegments(arr, segments);
    }

    private List<Segment> detectSegments(List<Integer> arr) {
        List<Segment> segments = new ArrayList<>();
        int n = arr.size();
        if (n == 0) {
            return segments;
        }

        int start = 0;
        while (start < n) {
            int end = start + 1;
            if (end < n && arr.get(start) > arr.get(end)) {
                // Decreasing segment
                while (end + 1 < n && arr.get(end) > arr.get(end + 1)) {
                    end++;
                }
                segments.add(new Segment(end, start)); // Store as (end, start) for backward iteration
            } else {
                // Increasing segment
                while (end + 1 < n && arr.get(end) <= arr.get(end + 1)) {
                    end++;
                }
                segments.add(new Segment(start, end));
            }
            start = end + 1;
        }
        return segments;
    }

    private void mergeSegments(List<Integer> arr, List<Segment> segments) {
        // A min-heap of HeapNode objects, compared by value.
        PriorityQueue<HeapNode> minHeap = new PriorityQueue<>(Comparator.comparingInt(node -> node.value));

        for (int i = 0; i < segments.size(); i++) {
            Segment seg = segments.get(i);
            int value = arr.get(seg.start);
            int direction = (seg.start <= seg.end) ? 1 : -1;
            minHeap.add(new HeapNode(value, i, seg.start, direction));
        }

        List<Integer> sortedArr = new ArrayList<>(arr.size());
        while (!minHeap.isEmpty()) {
            HeapNode node = minHeap.poll();
            sortedArr.add(node.value);

            int nextPos = node.currentPos + node.direction;
            Segment seg = segments.get(node.segIndex);

            if ((node.direction == 1 && nextPos <= seg.end) || (node.direction == -1 && nextPos >= seg.end)) {
                int nextValue = arr.get(nextPos);
                minHeap.add(new HeapNode(nextValue, node.segIndex, nextPos, node.direction));
            }
        }

        // Copy the sorted result back to the original list
        for (int i = 0; i < arr.size(); i++) {
            arr.set(i, sortedArr.get(i));
        }
    }

    // Example of usage
    public static void main(String[] args) {
        SegmentSort sorter = new SegmentSort();
        
        List<Integer> v = new ArrayList<>(List.of(5, 3, 2, 4, 6, 8, 7, 19, 10, 12, 13, 14, 17, 18));
        System.out.println("Original vector: " + v);
        sorter.sort(v);
        System.out.println("Sorted vector:   " + v);

        List<Integer> v2 = new ArrayList<>(List.of(9, 2, 3, 4, 5, 6, 7, 8, 1));
        System.out.println("\nOriginal vector: " + v2);
        sorter.sort(v2);
        System.out.println("Sorted vector:   " + v2);

        List<Integer> v3 = new ArrayList<>(List.of(1, 2, 3, 4, 5, 6, 7, 8, 9));
        System.out.println("\nOriginal vector: " + v3);
        sorter.sort(v3);
        System.out.println("Sorted vector:   " + v3);

        List<Integer> v4 = new ArrayList<>(List.of(9, 8, 7, 6, 5, 4, 3, 2, 1));
        System.out.println("\nOriginal vector: " + v4);
        sorter.sort(v4);
        System.out.println("Sorted vector:   " + v4);
    }
}

File: implementations/go/segmentsort.go
package main

import (
	"container/heap"
	"fmt"
)

// Segment stores the start and end indices of a detected segment.
type Segment struct {
	start, end int
}

// HeapNode represents an object in our min-heap.
type HeapNode struct {
	value      int // The value of the element
	segIndex   int // The segment this element belongs to
	currentPos int // The position of this element in the original array
	direction  int // The direction of iteration for this segment
}

// MinHeap implements heap.Interface for a slice of HeapNode pointers.
type MinHeap []*HeapNode

func (h MinHeap) Len() int { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i].value < h[j].value }
func (h MinHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }

// Push adds an element to the heap.
func (h *MinHeap) Push(x interface{}) {
	*h = append(*h, x.(*HeapNode))
}

// Pop removes and returns the minimum element from the heap.
func (h *MinHeap) Pop() interface{} {
	old := *h
	int n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// SegmentSort encapsulates the sorting logic.
type SegmentSort struct{}

// Sort sorts a slice of integers using the Segment Sort algorithm.
func (s *SegmentSort) Sort(arr []int) {
	n := len(arr)
	if n <= 1 {
		return
	}

	// Phase 1: Segment Detection
	segments := s.detectSegments(arr)

	// Phase 2: Heap Merging
	if len(segments) == 0 {
		return
	}

	s.mergeSegments(arr, segments)
}

func (s *SegmentSort) detectSegments(arr []int) []Segment {
	var segments []Segment
	n := len(arr)
	if n == 0 {
		return segments
	}

	start := 0
	for start < n {
		end := start + 1
		if end < n && arr[start] > arr[end] {
			// Decreasing segment
			for end+1 < n && arr[end] > arr[end+1] {
				end++
			}
			segments = append(segments, Segment{start: end, end: start})
		} else {
			// Increasing segment
			for end+1 < n && arr[end] <= arr[end+1] {
				end++
			}
			segments = append(segments, Segment{start: start, end: end})
		}
		start = end + 1
	}
	return segments
}

func (s *SegmentSort) mergeSegments(arr []int, segments []Segment) {
	minHeap := &MinHeap{}
	heap.Init(minHeap)

	for i, seg := range segments {
		value := arr[seg.start]
		direction := 1
		if seg.start > seg.end {
			direction = -1
		}
		heap.Push(minHeap, &HeapNode{
			value:      value,
			segIndex:   i,
			currentPos: seg.start,
			direction:  direction,
		})
	}

	sortedArr := make([]int, 0, len(arr))
	for minHeap.Len() > 0 {
		node := heap.Pop(minHeap).(*HeapNode)
		sortedArr = append(sortedArr, node.value)

		nextPos := node.currentPos + node.direction
		seg := segments[node.segIndex]

		if (node.direction == 1 && nextPos <= seg.end) || (node.direction == -1 && nextPos >= seg.end) {
			nextValue := arr[nextPos]
			heap.Push(minHeap, &HeapNode{
				value:      nextValue,
				segIndex:   node.segIndex,
				currentPos: nextPos,
				direction:  node.direction,
			})
		}
	}

	// Copy sorted result back
	copy(arr, sortedArr)
}

func main() {
	sorter := &SegmentSort{}

	v := []int{5, 3, 2, 4, 6, 8, 7, 19, 10, 12, 13, 14, 17, 18}
	fmt.Println("Original vector:", v)
	sorter.Sort(v)
	fmt.Println("Sorted vector:  ", v)

	v2 := []int{9, 2, 3, 4, 5, 6, 7, 8, 1}
	fmt.Println("\nOriginal vector:", v2)
	sorter.Sort(v2)
	fmt.Println("Sorted vector:  ", v2)

	v3 := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	fmt.Println("\nOriginal vector:", v3)
	sorter.Sort(v3)
	fmt.Println("Sorted vector:  ", v3)

	v4 := []int{9, 8, 7, 6, 5, 4, 3, 2, 1}
	fmt.Println("\nOriginal vector:", v4)
	sorter.Sort(v4)
	fmt.Println("Sorted vector:  ", v4)
}
